common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -237,6 +258,20 @@ namespace com.keyman.text {
           // For DOM-aware targets, this will trigger a DOM event page designers may listen for.
           outputTarget.doInputEvent();
         }
+
+        // The keyboard may want to take an action after all other keystroke processing is
+        // finished, for example to switch layers. This action may not have any output
+        // but may change system store or variable store values. Given this, we don't need to
+        // save anything about the post behavior, after finalizing it
+
+        // We need to tell the keyboard if the layer has been changed, either by a keyboard rule itself,
+        // or by the touch layout 'nextlayer' control.
+        this.keyboardProcessor.layerChangedStore.set(startingLayerId == this.keyboardProcessor.layerId ? '0' : '1');
+
+        let postRuleBehavior = this.keyboardProcessor.processPostKeystroke(keyEvent.device, outputTarget);
```

Comment:
Is `keyEvent.device` different from `this.device`? (l. 64)

Reply:
I don't think it is different, just another way of accessing the device.